controllers/workspace/predicates.go

```diff
@@ -47,3 +51,31 @@ var predicates = predicate.Funcs{
 	},
 	GenericFunc: func(_ event.GenericEvent) bool { return true },
 }
+
+var podPredicate = predicate.Funcs{
+	UpdateFunc: func(ev event.UpdateEvent) bool {
+		newObj, ok := ev.ObjectNew.(*corev1.Pod)
+		if !ok {
+			return true
+		}
+
+		oldObj, ok := ev.ObjectOld.(*corev1.Pod)
+		if !ok {
+			// Should never happen
+			return true
+		}
+
+		// If the dewworkspace label does not exist, do no reconcile
+		if _, ok := newObj.Labels[constants.DevWorkspaceIDLabel]; !ok {
+			return false
+		}
+
+		// Always reconcile if resource is deleted
+		if newObj.GetDeletionTimestamp() != nil {
+			return true
+		}
+
+		// Trigger a reconcile on failed workspaces if status is updated.
+		return !equality.Semantic.DeepEqual(oldObj.Status, newObj.Status)
```

Comment:
that's going to trigger reconcile for failed DW and as the result DW will be restarted.

I tried to rework it in the following way, but somebody else still trigger reconcile loop:
```go
func podPredicates(c client.Client) predicate.Funcs {
	return predicate.Funcs{
		UpdateFunc: func(ev event.UpdateEvent) bool {
			newObj, ok := ev.ObjectNew.(*corev1.Pod)
			if !ok {
				//that's no Pod. Let other predicates decide
				return true
			}

			oldObj, ok := ev.ObjectOld.(*corev1.Pod)
			if !ok {
				// Should never happen
				return true
			}

			// If the devworkspace label does not exist, do no reconcile
			var dwName string
			if dwName, ok = newObj.Labels[constants.DevWorkspaceNameLabel]; !ok {
				return false
			}

			// Do not trigger a reconcile on if status is not updated
			if equality.Semantic.DeepEqual(oldObj.Status, newObj.Status) {
				return false
			}
			//return false;
			workspace := &dw.DevWorkspace{}
			err := c.Get(context.TODO(), types.NamespacedName{Name: dwName, Namespace: newObj.GetNamespace()}, workspace)
			if err != nil {
				if !k8sErrors.IsNotFound(err) {
					//TODO log error
				}
				//Do not trigger if we failed to evaluate DW state. Otherwise failed workspaces will be restarted
				return false
			}

			noFailed := workspace.Status.Phase != dw.DevWorkspaceStatusFailed
			if noFailed {
				logf.Log.WithName("controller_devworkspace_config").Info("Pod: reconcile")
			} else {
				logf.Log.WithName("controller_devworkspace_config").Info("Pod: Do not reconcile pod for failed dw")
			}
			return noFailed
		},
	}
}
```

```
{"level":"info","ts":1617979194.2063107,"logger":"controller_devworkspace_config","msg":"Pod: Do not reconcile pod for failed dw"}
{"level":"info","ts":1617979194.2237866,"logger":"controllers.DevWorkspace","msg":"Error updating workspace status: Operation cannot be fulfilled on devworkspaces.workspace.devfile.io \"failure-test\": the object has been modified; please apply your changes to the latest version and try again","Request.Namespace":"devworkspace-controller","Request.Name":"failure-test","devworkspace_id":"workspace4531d661a2ff4323"}
{"level":"info","ts":1617979194.2240682,"logger":"controllers.DevWorkspace","msg":"Reconciling Workspace","Request.Namespace":"devworkspace-controller","Request.Name":"failure-test","devworkspace_id":"workspace4531d661a2ff4323"}

```

Reply:
That should be updated deployment status, since we scale it to 0. But how then it works in main branch.
We should rethink how we fail workspace start.
We should set started to false and force users to restart after failure, but it's not the best UX from operator perspective.
Or filter all event in the same way as I did pods above

Reply:
@sleshchenko I've updated the predicate to be almost the same, I've removed the part on getting the workspace though. Instead, went with updating the workspace.Spec.Started field and rely on the failWorkspace + stopWorkspace funcs. It seems to work! WDYT?

Reply:
I'm OK with that. I'll take a look changes.